### PR TITLE
docs(proposal): edges, buffers and buckets

### PR DESCRIPTION
**Pull request by whynowy**
Explain what this PR does.

### CODE REVIEW

File: docs/specifications/edges-buffers-buckets.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      to: sink
+```
+
+Each `edge` could have a name for internal usage, the naming convention is `{pipeline-name}-{from-vertex-name}-{to-vertex-name}`.
+
+## Buffers
+
+`Buffer` is `InterStepBuffer`. Each buffer has an owner, which is the vertex who reads from it. Each `udf` and `sink` vertex in a pipeline owns a group of partitioned buffers. Each buffer has a name with the naming convention `{pipeline-name}-{vertex-name}-{index}`, where the `index` is the partition index, starting from 0. This naming convention applies to the buffers of both map and reduce udf vertices.
+
+When multiple vertices connecting to the same vertex, if the to vertex is a map, the data from all the from vertices will be forwarded to the group of partitoned buffers round-robinly. If the to vertex is a reduce, the data from all the from vertices will be forwarded to the group of partitoned buffers based on the partitioning key.
```
Comment (whynowy) [inline docs/specifications/edges-buffers-buckets.md:55]: We will be able to easily support diamond dags.

File: docs/specifications/edges-buffers-buckets.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+A source vertex does not have any owned buffers. But a pipeline may have multiple source vertices, followed by one vertex. Same as above, if the following vertex is a map, the data from all the source vertices will be forwarded to the group of partitoned buffers round-robinly. If it is a reduce, the data from all the source vertices will be forwarded to the group of partitoned buffers based on the partitioning key.
```
Comment (whynowy) [inline docs/specifications/edges-buffers-buckets.md:57]: Be able to support multiple sources.

File: docs/specifications/edges-buffers-buckets.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+## Buckets
+
+`Bucket` is a K/V store (or a pair of stores) used for watermark propagation.
+
+There are 3 types of buckets in a pipeline:
+
+- `Edge Bucket`: used for edge watermark propagation. For Map vertices or Sink vertices, each of them has one bucket; for Reduce vertices, each of them has N buckets, where N is the partitions.
```
Comment (whynowy) [inline docs/specifications/edges-buffers-buckets.md:65]: Carefully review this: 

1. Use 1 bucket (2 K/V stores) for a Map with multi-partitioned buffers;
2. Use N buckets (2*N K/V stores) for a multi-partitioned reduce.

Are we able to do the same as Map for Reduce?

File: docs/specifications/edges-buffers-buckets.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      parallelism: 2
+    - from: compute-sum
+      to: sink
+```
+
+Each `edge` could have a name for internal usage, the naming convention is `{pipeline-name}-{from-vertex-name}-{to-vertex-name}`.
+
+## Buffers
+
+`Buffer` is `InterStepBuffer`. Each buffer has an owner, which is the vertex who reads from it. Each `udf` and `sink` vertex in a pipeline owns a group of partitioned buffers. Each buffer has a name with the naming convention `{pipeline-name}-{vertex-name}-{index}`, where the `index` is the partition index, starting from 0. This naming convention applies to the buffers of both map and reduce udf vertices.
```
Comment (whynowy) [inline docs/specifications/edges-buffers-buckets.md:53]: `edge.parallelism` should be moved to vertex.

File: docs/specifications/edges-buffers-buckets.md (hunk 1 above)
Comment (whynowy) [reply docs/specifications/edges-buffers-buckets.md:55]: `edge.parallelism` should be moved to vertex, otherwise it does not make sense for diamond dags.

**Review by vigith — COMMENTED** [review]
Discussed offline

**Comment by vigith** [comment]
* Buffer name will be just `{to}` for both Map and Reduce
* Buffer Bucket for WM and Metadata will be `{from}-{to}`, and there will be only one bucket for an edge, immaterial of how many partitions we will have, for both Map and Reduce.

**Comment by whynowy** [comment]
Updated.

**Review by vigith — APPROVED** [review]

File: docs/specifications/edges-buffers-buckets.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      to: out
+```
+
+Each `edge` could have a name for internal usage, the naming convention is `{pipeline-name}-{from-vertex-name}-{to-vertex-name}`.
+
+## Buffers
+
+`Buffer` is `InterStepBuffer`. Each buffer has an owner, which is the vertex who reads from it. Each `udf` and `sink` vertex in a pipeline owns a group of partitioned buffers. Each buffer has a name with the naming convention `{pipeline-name}-{vertex-name}-{index}`, where the `index` is the partition index, starting from 0. This naming convention applies to the buffers of both map and reduce udf vertices.
+
+When multiple vertices connecting to the same vertex, if the `to` vertex is a Map, the data from all the from vertices will be forwarded to the group of partitoned buffers round-robinly. If the `to` vertex is a Reduce, the data from all the from vertices will be forwarded to the group of partitoned buffers based on the partitioning key.
```
Comment (KeranYang) [inline docs/specifications/edges-buffers-buckets.md:59]: Before the proposal, if the target vertex is Map/Sink, there is one partition buffer, if Reduce, the number of partitions is `parallelism`. With this proposal, how do we determine how many partitioned buffers are allocated to the vertex?

Comment (KeranYang) [reply docs/specifications/edges-buffers-buckets.md:59]: Synced offline. I missed comments about making `parallelism` as vertex level attribute.

**Review by KeranYang — APPROVED** [review]

**Comment by whynowy** [comment]
@numaproj/numaflow-dev - Please review.

File: docs/specifications/edges-buffers-buckets.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+A Source vertex does not have any owned buffers. But a pipeline may have multiple Source vertices, followed by one vertex. Same as above, if the following vertex is a map, the data from all the Source vertices will be forwarded to the group of partitoned buffers round-robinly. If it is a reduce, the data from all the Source vertices will be forwarded to the group of partitoned buffers based on the partitioning key.
```
Comment (ashwinidulams) [inline docs/specifications/edges-buffers-buckets.md:61]: i see that we propose that map vertex could own multiple partitions. I understand from offline conversation that we dont necessarily intend to develop multi-stream readers and so therefore, will there be a logic to determine which partition needs to be allocated to which map pod?

Comment (ashwinidulams) [reply docs/specifications/edges-buffers-buckets.md:61]: if the above statements are true, then during autoscaling, we may end up having multiple streams processed by a single vertex? or is the expectation that there will be no more than 1 stream/partition per vertex-pod?

File: docs/specifications/edges-buffers-buckets.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+## Buckets
+
+`Bucket` is a K/V store (or a pair of stores) used for watermark propagation.
+
+There are 3 types of buckets in a pipeline:
+
+- `Edge Bucket`: Each edge has a bucket, used for edge watermark propagation, no matter if the vertex that the edge leads to is a Map or a Reduce. The naming convention of an edge bucket is `{pipeline-name}-{from-vertex-name}-{to-vertex-name}`.
```
Comment (ashwinidulams) [inline docs/specifications/edges-buffers-buckets.md:69]: so if i understand correctly, this bucket will contain information for all the partitions (streams) ?

File: docs/specifications/edges-buffers-buckets.md (hunk 6 above)
Comment (whynowy) [reply docs/specifications/edges-buffers-buckets.md:61]: The proposal for multi partitioned buffer is, one vertex pod will start N Goroutines (N=partitions) to read messages, each on is responsible for 1 partition.

Comment (ashwinidulams) [reply docs/specifications/edges-buffers-buckets.md:61]: makes sense

**Review by ashwinidulams — APPROVED** [review]

File: docs/specifications/edges-buffers-buckets.md (hunk 2 above)
Comment (whynowy) [reply docs/specifications/edges-buffers-buckets.md]: > here, partition refers to a stream (jetstream) as of today?

yes

File: docs/specifications/edges-buffers-buckets.md (hunk 3 above)
Comment (xdevxy) [reply docs/specifications/edges-buffers-buckets.md]: Just to be clear, for `Edge Bucket`, no matter how many edges the vertex has there would be one bucket for map, and N buckets for reduce?

Comment (whynowy) [reply docs/specifications/edges-buffers-buckets.md]: > Just to be clear, for `Edge Bucket`, no matter how many edges the vertex has there would be one bucket for map, and N buckets for reduce?

No, it's been updated to 1 bucket for both use cases.

**Review by xdevxy — COMMENTED** [review]
Just for my own understanding, in summary 1) each vertex owns the incoming buffer no matter how many incoming edges there are. 2) each edge will have one corresponding edge bucket. Is that correct?

**Comment by whynowy** [comment]
> Just for my own understanding, in summary 1) each vertex owns the incoming buffer no matter how many incoming edges there are. 2) each edge will have one corresponding edge bucket no matter how many edges there are between two vertices. Is that correct?

Good summary!

**Pull request merged**